### PR TITLE
implement filter and fix outdated docs

### PR DESCRIPTION
**Pull request by turboMaCk**
Hi! I would like to use this library but I'm missing filter function (which arguably could be worked around by using toList and fromList) but that seems a bit wasteful.

I also noticed some of the documentation is outdated since `Stack` now also contains the length in constructor so I made these updates as well.

### CODE REVIEW

File: src/Stack.elm [hunk 1]
```diff
@@ -62,27 +62,27 @@ toList (Stack _ xs) = xs
...
     The second element is a new `Stack` without the top of the stack
     or an empty `Stack` if the given `Stack` was empty.
 
-      Stack.pop Stack.empty     => (Nothing, Stack.empty) 
+      Stack.pop Stack.empty     => (Nothing, Stack.empty)
 
-      Stack.fromList [1, 2, 3]  
-        |> Stack.pop            => (Just 1, Stack [2, 3]) 
+      Stack.fromList [1, 2, 3]
+        |> Stack.pop            => (Just 1, Stack 2 [2, 3])
```
Comment (ntreu14) [inline src/Stack.elm:80]: Nice! Thanks for updating this to include the length, didn't even think to do this when I originally wrote the docs!

File: src/Stack.elm [hunk 2]
```diff
@@ -134,30 +134,54 @@ member sought (Stack _ xs) =
...
-      Stack.concat stacks      => Stack [1, 2, 3, 4, 5, 6]
+      Stack.concat stacks      => Stack 6 [1, 2, 3, 4, 5, 6]
 -}
 concat : List (Stack a) -> Stack a
-concat = List.foldr append empty 
+concat = List.foldr append empty
 
-{-| Given a function `(a -> b)` and a `Stack a` return a new 
+{-| Given a function `(a -> b)` and a `Stack a` return a new
     `Stack b` with the function applied to every element of the stack.
```
Comment (ntreu14) [inline src/Stack.elm]: I know this isn't really part of your PR, but can you add `Similar to List.map` here like you did for `filter`?

File: src/Stack.elm [hunk 3]
```diff
@@ -134,30 +134,54 @@ member sought (Stack _ xs) =
...
+      Stack.fromList [1, 2, 3]
+        |> Stack.map ((+) 1)      => Stack 3 [2, 3, 4]
 -}
 map : (a -> b) -> Stack a -> Stack b
 map f (Stack count xs) =
   List.map f xs
     |> Stack count
+
+
+{-| Filter items within stack, similar to `List.filter`
```
Comment (ntreu14) [inline src/Stack.elm]: I think to better match the verbiage for `map`, can we update this to be:

```
Give a predicate function `(a -> Bool)` and a `Stack a` return a new
`Stack a` with only the elements in which the predicate function returns `True`. Similar to `List.filter`. 
```

I also like the "Similar to `List`" inclusion. Let's add that to `map` too!

File: src/Stack.elm [hunk 4]
```diff
@@ -134,30 +134,54 @@ member sought (Stack _ xs) =
...
+
+      Stack.fromList [1, 2, 3]
+        |> Stack.filter ((<) 1)      => Stack 2 [2, 3]
+
+-}
+filter : (a -> Bool) -> Stack a -> Stack a
+filter f (Stack _ xs) =
```
Comment (ntreu14) [inline src/Stack.elm:177]: I like this implementation, but there is an arguably more simple one:
```elm
filter : (a -> Bool) -> Stack a -> Stack a
filter predicate (Stack _ xs) =
  let
    newList = List.filter predicate xs 
  in
  Stack (List.length newList) newList
```

Did you consider performance in your implementation? It does look like [List.filter](https://github.com/elm/core/blob/master/src/List.elm#L213) and [List.length](https://github.com/elm/core/blob/master/src/List.elm#L250) use `foldr`, so your solution removes the need to do that twice which I like.

What are your thoughts?

**Review by ntreu14 — CHANGES_REQUESTED** [review]
Looks good so far, thanks for your contribution! 

I have a couple of comments and please add unit tests for `filter`.

File: src/Stack.elm (hunk 4 above)
Comment (turboMaCk) [reply src/Stack.elm:177]: I did not try to measure the real performance implications. Both `length` and `filter` has to visit every item in the list at least once*. With folder both of these are done together so in theory it should be faster. But I believe both implementation would be O(n) so scalability would not be concern in either case.

I've went for the version which seems algorithmically more optimal because it was still faster to implement optimal version than trying to measure cost of `List.filter` + `List.length`.

---
*) filter also produces with same order as `filter (\_ -> True) xs == xs` so like a foldr it must deal with this ordering while traversing so I guess it technically is not just single visit per item.

File: src/Stack.elm (hunk 1 above)
Comment (turboMaCk) [reply src/Stack.elm:80]: ah I assumed the count must been added later on and docs were not updated. Anyway with this change it should at least match what repl would print if you would copy the code there. There are also ways to run these examples as tests btw.

File: tests/Tests.elm [hunk 5]
```diff
@@ -19,27 +19,27 @@ both : (a -> b) -> (a, a) -> (b, b)
...
-    [ describe "empty" 
-        [ test "has a size of 0" <| 
+tests =
+  describe "Stack tests"
+    [ describe "empty"
+        [ test "has a size of 0" <|
             \() ->
                 Stack.size Stack.empty |> Expect.equal 0
-        
+
```
Comment (turboMaCk) [inline tests/Tests.elm:28]: sorry for these changes in diff. I have my editor configured to automatically strip trailing white spaces and enforce unix style file ends (made to not mess with C preprocessor). Let me know if you want me to revert these changes.

File: tests/Tests.elm [hunk 6]
```diff
@@ -221,11 +221,23 @@ tests =
...
           ]
-    ]
\ No newline at end of file
+    , describe "filter"
+          [ fuzz (list int) "filter (always True) satisfies identity law" <|
+              \xs ->
+                  Stack.fromList xs
+                    |> Stack.filter (always True)
+                    |> Expect.equal (Stack.fromList xs)
+          , test "result stack doesn't contain items which doesn't satisfy the predicate" <|
```
Comment (turboMaCk) [inline tests/Tests.elm]:
```suggestion
          , test "result stack doesn't contain items which don't satisfy the predicate" <|
```

File: tests/Tests.elm [hunk 7]
```diff
@@ -221,11 +221,23 @@ tests =
...
+
+          , test "mapping an empty stack should result in an empty stack" <|
               \() ->
                   Expect.equal
                     Stack.empty
                     (Stack.map (always "foobar") Stack.empty)
           ]
-    ]
\ No newline at end of file
+    , describe "filter"
```
Comment (ntreu14) [inline tests/Tests.elm:231]: Thanks for adding these! I know that this is might be overkill, but let's add a test for filtering an empty stack, should also be an empty stack.

File: tests/Tests.elm (hunk 5 above)
Comment (ntreu14) [reply tests/Tests.elm:28]: Not an issue!

File: src/Stack.elm (hunk 4 above)
Comment (ntreu14) [reply src/Stack.elm:177]: I'm cool with this implementation, I just wanted to hear your reasoning behind it!

File: src/Stack.elm [hunk 8]
```diff
@@ -134,30 +134,58 @@ member sought (Stack _ xs) =
...
+      Stack.fromList [1, 2, 3]
+        |> Stack.map String.fromInt      => Stack 3 ["1", "2", "3"]
 -}
 map : (a -> b) -> Stack a -> Stack b
 map f (Stack count xs) =
   List.map f xs
     |> Stack count
+
+
+{-| Give a predicate function `(a -> Bool)` and a `Stack a` return a new
```
Comment (ntreu14) [inline src/Stack.elm]:
```suggestion
{-| Given a predicate function `(a -> Bool)` and a `Stack a` return a new
```

Comment (ntreu14) [reply src/Stack.elm]: Sorry, I think that I had a typo in my original comment.

**Review by ntreu14 — COMMENTED** [review]
Looking really good. Let's add one more test for filtering an empty stack and I'll merge it in and get a release out.

**Review by ntreu14 — CHANGES_REQUESTED** [review]
Just need a test for filter an empty stack is an empty stack and I'll merge this and get a release out!

**Review by ntreu14 — APPROVED** [review]

**Comment by ntreu14** [comment]
@turboMaCk A new package version has been released! If you want to use it, please use version `1.1.0`.
